File: shells/android/app/src/main/java/dev/shellwright/shell/config/FastConfigReader.kt
package dev.shellwright.shell.config

/**
 * Phase one of the two-phase config parse.
 *
 * ⚠️ Parsing a 40 KB config with a reflective or generated JSON parser costs
 * real milliseconds on a cold JIT, on a budget device, on the main thread,
 * before anything is on screen. The startup budget is 300 ms to first frame
 * (`03_TEST_STRATEGY.md` §12) and this must not eat a tenth of it.
 *
 * So the first frame is drawn from this: a hand-written scanner that pulls out
 * only the handful of values needed to paint the native skeleton — theme
 * colours, tab labels, the start URL. The full [ShellConfig] is parsed on a
 * background dispatcher afterwards and everything else waits for it.
 *
 * This reads a *flat subset* deliberately. It is not a JSON parser and must not
 * grow into one; if a field is not needed for the first frame, it belongs in
 * phase two.
 *
 * Budget: under 5 ms on a mid-range device (`TC-S02-PRF-003`).
 */
public object FastConfigReader {

    /** The values needed to draw the native skeleton before any web content. */
    public data class FirstFrame(
        val appName: String,
        val initialUrl: String,
        val splashBackground: String,
        val themePrimary: String,
        val themeNavBar: String,
        val themeTabBar: String,
        val statusBarStyle: String,
        val topBarEnabled: Boolean,
        val tabBarEnabled: Boolean,
        val tabLabels: List<String>,
    )

    private const val DEFAULT_WHITE = "#FFFFFF"

    /**
     * Scans the raw config for first-frame values.
     *
     * Never throws: a malformed config still has to draw something, and phase
     * two reports the real error. Every field falls back to the schema default.
     */
    public fun read(json: String): FirstFrame = FirstFrame(
        appName = stringAfter(json, "\"name\"") ?: "",
        initialUrl = stringAfter(json, "\"initialUrl\"") ?: "",
        splashBackground = stringAfter(json, "\"backgroundColor\"") ?: DEFAULT_WHITE,
        themePrimary = stringAfter(json, "\"primary\"") ?: "#2563EB",
        themeNavBar = stringAfter(json, "\"navBar\"") ?: DEFAULT_WHITE,
        themeTabBar = stringAfter(json, "\"tabBar\"") ?: DEFAULT_WHITE,
        statusBarStyle = stringAfter(json, "\"statusBar\"") ?: "dark-content",
        topBarEnabled = enabledUnder(json, "\"topBar\"", default = true),
        tabBarEnabled = enabledUnder(json, "\"tabBar\"", default = false),
        tabLabels = tabLabels(json),
    )

    /**
     * The string value following the first occurrence of [key].
     *
     * Returns null when the key is absent, or when its value is not a plain
     * string — `"tabBar": { … }` must not be mistaken for a colour.
     */
    private fun stringAfter(json: String, key: String): String? {
        val keyAt = json.indexOf(key)
        if (keyAt < 0) return null

        val colon = json.indexOf(':', keyAt + key.length)
        if (colon < 0) return null

        val openQuote = json.indexOfFirst(colon + 1) { it == '"' || it == '{' || it == '[' }
        if (openQuote < 0 || json[openQuote] != '"') return null

        return readString(json, openQuote)
    }

    /** Reads a JSON string starting at its opening quote, honouring escapes. */
    private fun readString(json: String, openQuote: Int): String? {
        val builder = StringBuilder()
        var i = openQuote + 1

        while (i < json.length) {
            when (val ch = json[i]) {
                '"' -> return builder.toString()
                '\\' -> {
                    if (i + 1 >= json.length) return null
                    // Phase one only reads names, URLs, colours, and labels.
                    // Unicode escapes are left to phase two rather than
                    // reimplementing them here.
                    val escaped = json[i + 1]
                    if (escaped == 'u') return null
                    builder.append(unescape(escaped))
                    i += 2
                }
                else -> {
                    builder.append(ch)
                    i++
                }
            }
        }
        return null
    }

    private fun unescape(escaped: Char): Char = when (escaped) {
        'n' -> '\n'
        't' -> '\t'
        'r' -> '\r'
        'b' -> '\b'
        else -> escaped
    }

    /** Whether the object introduced by [key] has `"enabled": true`. */
    private fun enabledUnder(json: String, key: String, default: Boolean): Boolean {
        val keyAt = json.indexOf(key)
        if (keyAt < 0) return default

        val enabledAt = json.indexOf("\"enabled\"", keyAt)
        if (enabledAt < 0) return default

        // Only trust an "enabled" that belongs to this object: anything past the
        // next object key at the same nesting level is somebody else's.
        val valueAt = json.indexOf(':', enabledAt)
        if (valueAt < 0) return default

        val trueAt = json.indexOf("true", valueAt)
        val falseAt = json.indexOf("false", valueAt)
        return when {
            trueAt in 0..(valueAt + 4) -> true
            falseAt in 0..(valueAt + 4) -> false
            else -> default
        }
    }

    /** Tab labels in document order, for painting the bar before web content. */
    private fun tabLabels(json: String): List<String> {
        val tabBarAt = json.indexOf("\"tabBar\"")
        if (tabBarAt < 0) return emptyList()

        val itemsAt = json.indexOf("\"items\"", tabBarAt)
        if (itemsAt < 0) return emptyList()

        val labels = mutableListOf<String>()
        var cursor = itemsAt
        val end = json.indexOf(']', itemsAt).takeIf { it >= 0 } ?: json.length

        while (labels.size < MAX_TABS) {
            val labelAt = json.indexOf("\"label\"", cursor)
            if (labelAt < 0 || labelAt > end) break

            val colon = json.indexOf(':', labelAt)
            if (colon < 0) break

            val valueAt = json.indexOfFirst(colon + 1) { !it.isWhitespace() }
            if (valueAt < 0) break

            // A translated label is an object; phase two resolves it properly.
            if (json[valueAt] == '"') {
                readString(json, valueAt)?.let(labels::add)
            } else {
                labels.add("")
            }
            cursor = valueAt + 1
        }
        return labels
    }

    /** iOS shows five; more than eight is beyond anything the schema allows. */
    private const val MAX_TABS = 8

    private inline fun String.indexOfFirst(from: Int, predicate: (Char) -> Boolean): Int {
        for (i in from until length) {
            if (predicate(this[i])) return i
        }
        return -1
    }
}

File: shells/ios/Sources/ShellCore/FastConfigReader.swift
import Foundation

/// Phase one of the two-phase config parse.
///
/// - Important: Decoding a 40 KB config through `Codable` costs real
///   milliseconds on a cold launch, on the main thread, before anything is on
///   screen. The startup budget is 300 ms to first frame
///   (`03_TEST_STRATEGY.md` §12) and this must not eat a tenth of it.
///
/// So the first frame is drawn from this: a hand-written scanner that pulls out
/// only the handful of values needed to paint the native skeleton — theme
/// colours, tab labels, the start URL. The full ``ShellConfig`` is decoded on a
/// background queue afterwards and everything else waits for it.
///
/// This reads a *flat subset* deliberately. It is not a JSON parser and must
/// not grow into one; if a field is not needed for the first frame, it belongs
/// in phase two.
///
/// Mirrors the Kotlin `FastConfigReader` field for field.
public enum FastConfigReader {
    /// The values needed to draw the native skeleton before any web content.
    public struct FirstFrame: Sendable, Equatable {
        public let appName: String
        public let initialUrl: String
        public let splashBackground: String
        public let themePrimary: String
        public let themeNavBar: String
        public let themeTabBar: String
        public let statusBarStyle: String
        public let topBarEnabled: Bool
        public let tabBarEnabled: Bool
        public let tabLabels: [String]
    }

    private static let defaultWhite = "#FFFFFF"
    private static let defaultPrimary = "#2563EB"

    /// iOS shows five tabs; more than eight is beyond anything the schema allows.
    private static let maxTabs = 8

    /// Scans the raw config for first-frame values.
    ///
    /// Never throws: a malformed config still has to draw something, and phase
    /// two reports the real error. Every field falls back to the schema default.
    public static func read(_ json: String) -> FirstFrame {
        let scalars = Array(json.unicodeScalars)

        return FirstFrame(
            appName: string(after: "\"name\"", in: scalars) ?? "",
            initialUrl: string(after: "\"initialUrl\"", in: scalars) ?? "",
            splashBackground: string(after: "\"backgroundColor\"", in: scalars) ?? defaultWhite,
            themePrimary: string(after: "\"primary\"", in: scalars) ?? defaultPrimary,
            themeNavBar: string(after: "\"navBar\"", in: scalars) ?? defaultWhite,
            themeTabBar: string(after: "\"tabBar\"", in: scalars) ?? defaultWhite,
            statusBarStyle: string(after: "\"statusBar\"", in: scalars) ?? "dark-content",
            topBarEnabled: enabled(under: "\"topBar\"", in: scalars, default: true),
            tabBarEnabled: enabled(under: "\"tabBar\"", in: scalars, default: false),
            tabLabels: tabLabels(in: scalars)
        )
    }

    /// The string value following the first occurrence of `key`.
    ///
    /// Returns nil when the key is absent, or when its value is not a plain
    /// string — `"tabBar": { … }` must not be mistaken for a colour.
    private static func string(after key: String, in scalars: [Unicode.Scalar]) -> String? {
        guard let keyAt = index(of: key, in: scalars) else { return nil }
        guard let colon = index(of: ":", in: scalars, from: keyAt + key.unicodeScalars.count)
        else { return nil }

        guard let valueAt = firstIndex(from: colon + 1, in: scalars, where: { !isWhitespace($0) })
        else { return nil }

        guard scalars[valueAt] == "\"" else { return nil }
        return readString(from: valueAt, in: scalars)
    }

    /// Reads a JSON string starting at its opening quote, honouring escapes.
    private static func readString(from openQuote: Int, in scalars: [Unicode.Scalar]) -> String? {
        var result = String.UnicodeScalarView()
        var i = openQuote + 1

        while i < scalars.count {
            let scalar = scalars[i]

            if scalar == "\"" {
                return String(result)
            }

            if scalar == "\\" {
                guard i + 1 < scalars.count else { return nil }
                let escaped = scalars[i + 1]
                // Phase one only reads names, URLs, colours, and labels.
                // Unicode escapes are left to phase two rather than
                // reimplementing them here.
                if escaped == "u" { return nil }
                result.append(unescape(escaped))
                i += 2
                continue
            }

            result.append(scalar)
            i += 1
        }

        return nil
    }

    private static func unescape(_ escaped: Unicode.Scalar) -> Unicode.Scalar {
        switch escaped {
        case "n": "\n"
        case "t": "\t"
        case "r": "\r"
        default: escaped
        }
    }

    /// Whether the object introduced by `key` has `"enabled": true`.
    private static func enabled(
        under key: String,
        in scalars: [Unicode.Scalar],
        default fallback: Bool
    ) -> Bool {
        guard let keyAt = index(of: key, in: scalars) else { return fallback }
        guard let enabledAt = index(of: "\"enabled\"", in: scalars, from: keyAt) else { return fallback }
        guard let colon = index(of: ":", in: scalars, from: enabledAt) else { return fallback }

        // Only trust a value that immediately follows the key, so a later
        // object's "enabled" is not mistaken for this one's.
        guard let valueAt = firstIndex(from: colon + 1, in: scalars, where: { !isWhitespace($0) })
        else { return fallback }

        if matches("true", at: valueAt, in: scalars) { return true }
        if matches("false", at: valueAt, in: scalars) { return false }
        return fallback
    }

    /// Tab labels in document order, for painting the bar before web content.
    private static func tabLabels(in scalars: [Unicode.Scalar]) -> [String] {
        guard let tabBarAt = index(of: "\"tabBar\"", in: scalars) else { return [] }
        guard let itemsAt = index(of: "\"items\"", in: scalars, from: tabBarAt) else { return [] }

        let end = index(of: "]", in: scalars, from: itemsAt) ?? scalars.count
        var labels: [String] = []
        var cursor = itemsAt

        while labels.count < maxTabs {
            guard let labelAt = index(of: "\"label\"", in: scalars, from: cursor), labelAt < end
            else { break }
            guard let colon = index(of: ":", in: scalars, from: labelAt) else { break }
            guard let valueAt = firstIndex(from: colon + 1, in: scalars, where: { !isWhitespace($0) })
            else { break }

            // A translated label is an object; phase two resolves it properly.
            if scalars[valueAt] == "\"" {
                if let label = readString(from: valueAt, in: scalars) {
                    labels.append(label)
                }
            } else {
                labels.append("")
            }

            cursor = valueAt + 1
        }

        return labels
    }

    // MARK: - Scalar scanning

    private static func index(
        of needle: String,
        in scalars: [Unicode.Scalar],
        from start: Int = 0
    ) -> Int? {
        let pattern = Array(needle.unicodeScalars)
        guard !pattern.isEmpty, start >= 0 else { return nil }
        guard scalars.count >= pattern.count else { return nil }

        for i in start...(scalars.count - pattern.count) where matchesPattern(pattern, at: i, in: scalars) {
            return i
        }
        return nil
    }

    private static func matches(_ needle: String, at index: Int, in scalars: [Unicode.Scalar]) -> Bool {
        matchesPattern(Array(needle.unicodeScalars), at: index, in: scalars)
    }

    private static func matchesPattern(
        _ pattern: [Unicode.Scalar],
        at index: Int,
        in scalars: [Unicode.Scalar]
    ) -> Bool {
        guard index >= 0, index + pattern.count <= scalars.count else { return false }
        for offset in 0..<pattern.count where scalars[index + offset] != pattern[offset] {
            return false
        }
        return true
    }

    private static func firstIndex(
        from start: Int,
        in scalars: [Unicode.Scalar],
        where predicate: (Unicode.Scalar) -> Bool
    ) -> Int? {
        guard start >= 0 else { return nil }
        for i in start..<scalars.count where predicate(scalars[i]) {
            return i
        }
        return nil
    }

    private static func isWhitespace(_ scalar: Unicode.Scalar) -> Bool {
        scalar == " " || scalar == "\n" || scalar == "\r" || scalar == "\t"
    }
}
